---
title: "Genotyping Accuracy Assessments"
output: 
  html_notebook:
    toc: true
---


This document explores the data and meta-data we have to make sure that everything
is looking like it should, and to do a few summaries.

To start off with, let's load data and libs:
```{r load-stuff}
library(tidyverse)
library(broom)

meta <- readRDS("../data/processed/meta-data-tibble.rds")
genos <- readRDS("../extdata/processed/called_genos_na_explicit.rds") %>%
  filter(NMFS_DNA_ID %in% meta$NMFS_DNA_ID)  # drop those we don't have meta data for
samples <- readRDS("../data/processed/sample-sheet-tibble.rds") %>%
  filter(NMFS_DNA_ID %in% meta$NMFS_DNA_ID)
```


## Regenotyped Individuals

Let's first tally up every NMFS_DNA_ID that was genotyped more times than once:
```{r tally-dupies}
regeno_tbl <- samples %>% 
  group_by(NMFS_DNA_ID) %>% 
  tally() %>% 
  filter(n>1) %>% 
  arrange(desc(n))
head(regeno_tbl)
```
This shows us that no one was genotyped more than twice.  There were `r nrow(regeno_tbl)` 
fish that were genotyped twice.  We want to compare the genotype calls between them all.

The following code puts the recalled genotypes into a sorted order from which
we should be able to determine easily if they are re-scored concordantly or not.  

But, also,  lots of individuals appear to not have any called genotypes on some runs.  Like perhaps they
were such a loss that the indivdual id never even made it into the genotypes.  Individuals x locus
combos in which all the genotypes are not called or not helpful to us here, so let us count up 
the number of called alleles in each pair of genotypings.  It must be four to be useful.  We add a column
that denotes that.
```{r find-twicers}
twice_called <- genos %>%
  filter(NMFS_DNA_ID %in% regeno_tbl$NMFS_DNA_ID) %>%
  arrange(NMFS_DNA_ID, locus, gtseq_run, allele) %>%
  group_by(NMFS_DNA_ID, locus) %>%
  mutate(gene_copies_called = sum(!is.na(allele))) %>%
  ungroup()
```


```{r filter-twicers}
cop_called_summary <- twice_called %>%
  group_by(NMFS_DNA_ID, locus) %>%
  mutate(gene_copies_called = sum(!is.na(allele)))
```

And now we add another column that includes, for each four rows or an individual
at a locus, either a TRUE the genotype is concordant, FALSE if not, and NA if
either of the times the individual was genotyped it came out NA.  
```{r concordo}
# make a function that takes the 4-vector of alleles sorted appropriately
concordo_func <- function(a) {
  if(any(is.na(a))) {
    ret <- NA
  } else if(all(a[1:2]==a[3:4])) {
    ret <- TRUE
  } else {
    ret <- FALSE
  }
  ret
}
# then apply it
#twice_called %>%
#  group_by(NMFS_DNA_ID, locus, gtseq_run) %>%
#  mutate(concordant = concordo_func(allele))
```


## Homozygosity as a function of read depth

Having seen a few discordant re-genotyped calls, it seems to me that it is often
occurring when a homozygote has been erroneously called when the read depth is low
(i.e. around 10 or so).  One thing I can do to investigate this is see if there is 
an evident trend at each locus---one in which an individual is more likely to be called
homozygous at lower read depth.  If there is, that would indicate that genotyping errors
occur more frequently at low read depth. (We might even be able to use this to estimate
the rate at which such genotyping error occur.)

### Starting with adult kelp rockfish

We are going to focus first on just the known kelp rockfish adults so that
we don't find a trend which is confounded by read depth differences and heterozygosity
differences in different species. This next bit of code picks them out and then creates
a column that says whether individuals are homozygous or not, and retains the total
read depth in each case.  
```{r get-adult-kelps}
old_kelps <- left_join(genos, meta %>% select(NMFS_DNA_ID, SPECIES, REPORTED_LIFE_STAGE)) %>%
  filter(SPECIES == "atrovirens", REPORTED_LIFE_STAGE == "ADULT") 

ok_homoz <- old_kelps %>%
  filter(!is.na(allele)) %>%
  group_by(gtseq_run, id, NMFS_DNA_ID, locus) %>%
  summarise(homozyg = allele[1] == allele[2],
            total_depth = {if(allele[1] == allele[2]){depth[1]} else sum(depth)},
            bal = min(allele.balance))

# and count up the number of non-NA genotypes at each locus
num_nonNA <- ok_homoz %>%
  group_by(locus, homozyg) %>%
  tally() %>%
  group_by(locus) %>%
  summarise(str = paste("Hom: ", n[homozyg==TRUE], " Het: ", n[homozyg==FALSE]), sep = "")
```
And now make a plot.  I can't get it to scale correctly as a PNG in the notebook so I 
just make a PDF of it here.
```{r plot-homoz, fig.width=12, fig.height=40}
g <- ggplot(ok_homoz, aes(x = total_depth)) +
  geom_density(aes(fill = homozyg), alpha = 0.5) +
  facet_wrap(~ locus, ncol = 4, scales = "free") +
  geom_text(data = num_nonNA, mapping = aes(label = str), x = 0, y = 0, hjust = 0, vjust = 0, colour = "white")

ggsave(g, file = "pdf_figs/kelp_adult_hom_hz_densities.pdf", width = 16, height = 40)
```
You can download that figure from [here](https://github.com/eriqande/nsf-rockfish-parentage/blob/master/Rmd/pdf_figs/kelp_adult_hom_hz_densities.pdf).

So, there are a few loci that give us some pause there.  In particular, let's see what is going on with 
`Plate_2_G06_Sat_EW987118_consensus`.  We can do that by ordering by read depth to see if we note an unexpected
trend of homozygotes at lower read depth:
```{r check-a-loc}
ok_homoz %>%
  filter(locus == "Plate_2_G06_Sat_EW987118_consensus") %>%
  arrange(total_depth) %>%
  filter(total_depth < 50)
```
That actually doesn't look as bad as it could.  The allele balances are pretty low on the hets, but there are some
hets at low read depth.  it is not totally ugly.  Cool.  

Let's look at one more, `tag_id_2319`:
```{r check-another-loc}
ok_homoz %>%
  filter(locus == "tag_id_2319") %>%
  arrange(total_depth) %>%
  filter(total_depth < 50)
```
That is not a huge smoking gun, of any sort.  So, this is looking pretty good.  Nice.


## Snuffing around for null alleles

One of the big advantages of doing amplicon sequencing, I have always thought, is that it should
be less subject to allelic dropout than methods like rapture or ddRAD-DART that depend
on a digestion step that is very specific to a certain enzyme recognition sequence.
Given that our primers amplify our loci in so many species, it is unlikely that they
will be subject to allelel-specific dropout, but it would be neat to have a look at that.

To do so we could model the total depth as a function of the locus, gtseq_run, and the id of the 
individual (I think it would be interesting to see how well we can predict total depth on the
basis of this, anyway) and then see if there are two clusters of homozygotes, one with half
as many reads as expected.  Could be a long shot, but we should check it out.  

I started with a simple linear model. But, because we have 
censored all the read depths less than 10, this is not right.  It runs, eventually, and
gives reasonable results, but is messy to deal with.  
original data for now.  
```{r depth-lm, eval=FALSE}
dfit <- lm(total_depth ~ locus + gtseq_run + id, data = ok_homoz)
```
We are going to leave that here for now.  