---
title: "Doing Parentage on the Kelp Rockfish"
output: 
  html_notebook:
    toc: true
---


This is a first stab at parentage on kelp rockfish.  In a previous document, I assumed that
all the adult kelp rockfish were correctly ID-ed and I used them as a baseline to do GSI of all
the juvenile samples that could have some kelp rockfish in them.  I save those results and use
that here...


To start off with, let's load data and libs:
```{r load-stuff}
library(gRbase)  
library(RBGL)  # this one comes from BioConductor: source("https://bioconductor.org/biocLite.R"); biocLite("RBGL")

library(tidyverse)
library(CKMRsim)
library(stringr)


meta <- readRDS("../data/processed/meta-data-tibble.rds")
genos <- readRDS("../extdata/processed/genos-aggregated-and-no-hi-missers.rds")
species <- readRDS("../data/processed/juvie-gsi-sim-assignments.rds")   # gsi assignments of juvenile fish
```


## Selecting our fish and designating allele nomenclature

We gotta pick out the _S. atrovirens_.  For adults that is pretty much already done, just by who they
are, but we must use the GSI assignments for the juveniles.
```{r define-kelps}
kelp_ad <- meta %>% 
  filter(SPECIES == "atrovirens",
         REPORTED_LIFE_STAGE == "ADULT") %>%
  select(NMFS_DNA_ID) %>% unlist() %>% unname()

kelp_juv <- species %>%
  filter(species == "atrovirens",
         score > 99.9) %>%
  select(NMFS_DNA_ID) %>% unlist() %>% unname

```

With those we can just filter down the genos to the ones that we want, and then we
can get it into the format required for CKMR.
```{r kelp-genos}
kelp_genos <- genos %>%
  filter(NMFS_DNA_ID %in% c(kelp_ad, kelp_juv))

# we will use this some more
kg2 <- kelp_genos %>% 
  select(NMFS_DNA_ID, locus, allele) %>%
  mutate(Chrom = "GTseq") %>% 
  mutate(Pos = as.integer(factor(locus, levels = unique(locus)))) %>%
  rename(Locus = locus,
         Allele = allele) %>%
  select(NMFS_DNA_ID, Chrom, Locus, Pos, Allele) %>%
  ungroup()

# get the allele freqs
kg_ckmr_markers <- kg2 %>%
  filter(!is.na(Allele)) %>% # it is vital to filter out the NAs at this stage
  group_by(Chrom, Locus, Pos, Allele) %>%
  summarise(counts = n()) %>%
  group_by(Locus, Pos) %>%
  mutate(Freq = counts / sum(counts)) %>%
  select(-counts) %>%
  mutate(AlleIdx = 1,
         LocIdx = 1) %>%
  reindex_markers(.)
```
Great! We have a total `r nrow(kg_ckmr_markers)` alleles in this data set.


## Doing CKMR simulations to assess power

Really simply let's just do this to see what our power for parentage ought to look like.
```{r CKMR-create}
CK <- create_ckmr(kg_ckmr_markers, kappa_matrix = kappas[c("PO", "FS", "HS", "U"), ])

# see what it tells us when we print it:
CK
```

Then do the simulations to see what the false positive rates are for finding parents when the truth
is U or FS:
```{r CKMR-sim-and-sample}
Qs <- simulate_Qij(C = CK, froms = c("PO", "FS", "HS", "U"), tos = c("PO", "FS", "HS", "U"), reps = 10^4)

# then do the importance sampling to get the FPRs
mc_sample_simple(Qs, nu = "PO", de = c("U", "FS"), tr = c("U", "FS"), method = "both")
```
We can come back to that later to see about finding Full Sibs.

## Making genotype matrices
```{r}
kelp_haps <- kelp_genos %>%
  filter(!is.na(allele)) %>%  # once again, it is critical to remove these at this point
  select(NMFS_DNA_ID, locus, gene_copy, allele) %>%
  rename(Locus = locus, Allele = allele)

kelp_idx_frame <- kg_ckmr_markers %>%
  select(Locus, Allele, LocIdx, AlleIdx) %>%
  group_by(Locus) %>%
  mutate(NumA = n()) %>%  # get the number of alleles at each locus
  ungroup() %>%
  left_join(kelp_haps, .)  %>%  # join the alle_idx's onto the actual genotype data
  select(NMFS_DNA_ID, Locus, gene_copy, LocIdx, NumA, AlleIdx) %>%
  spread(key = gene_copy, value = AlleIdx) %>%
  mutate(GenoIdx = index_ab(a = `1`, b = `2`, A = NumA))

# make a matrix of genotype integers 
wide_kelp <- kelp_idx_frame %>%
  select(NMFS_DNA_ID, LocIdx, GenoIdx) %>%
  spread(data = ., key = LocIdx, value = GenoIdx)
```

Don't forget to set NA's to 0, and then decrease each value by 1:
```{r make-mat}
rocky_mat <- as.matrix(wide_kelp[, -1])
rownames(rocky_mat) <- wide_kelp$NMFS_DNA_ID
rocky_mat[is.na(rocky_mat)] <- 0
rocky_mat <- rocky_mat - 1
storage.mode(rocky_mat) <-  "integer"
```

## Looking for duplicated samples

We can quickly look through rocky_mat for pairs of indivs with lots of matching genotypes.
```{r check-for-dupes}
matchers <- pairwise_geno_id(S = rocky_mat, max_miss = 12) %>%
  arrange(num_mismatch)
matchers
```
We need to deal with these somehow.  I am going to say that anyone with 3 or fewer mismatches is certainly 
the same individual, and we are going to take just one from amongst those pairs/groups.  We use a hasty method:
we toss out all the individuals that appear in both ind1 and ind2, and then all the remaining ind2's.  This 
might throw out one more than we need to, but it is what we do for now.
```{r toss-matchers}
def_same <- matchers %>%
  filter(num_mismatch <= 3)

toss_these <- c(intersect(def_same$ind1, def_same$ind2), def_same$ind2) %>%
  unique()

rocky_mat_tossed <- rocky_mat[-toss_these,]
```

Now we have to break that into two matrices: one for the adults and one for the kids.
This should not be too hard.  We have to deal with the fact that not all of the adults in the meta data
had few enough missing genotypes to be included.
```{r break-stage}
kelp_ad_kept <- kelp_ad[kelp_ad %in% rownames(rocky_mat_tossed)]
kelp_juv_kept <- kelp_juv[kelp_juv %in% rownames(rocky_mat_tossed)]

adult_mat <- rocky_mat_tossed[kelp_ad_kept, ]
juvie_mat <- rocky_mat_tossed[kelp_juv_kept, ]
```

## Doing the comparisons

We need to flatten the CKMR object appropriately
```{r flatten}
# flatten those out
po_flat <- flatten_ckmr(CK, "PO")
unrel_flat <- flatten_ckmr(CK, "U")

# then compute the log-likelihoods for the parent offspring relationship
po_logl_flat <- po_flat
po_logl_flat$probs <- log(po_flat$probs / unrel_flat$probs)

```


Now we are ready to do these!
```{r do-comps}
idx <- 1:nrow(juvie_mat)
names(idx) <- idx
comps <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = adult_mat, T = juvie_mat, t = i, values = po_logl_flat$probs, nGenos = po_logl_flat$nGenos, Starts = po_logl_flat$base0_locus_starts)
    #tmp[rev(top_index(tmp$value, 5)), ]  # just take the top 5 from each
    }) %>%
  dplyr::bind_rows(.id = "offspring") %>%
  dplyr::tbl_df() 

# have a look at that
result <- arrange(comps, desc(value)) %>%
  mutate(number = 1:nrow(.)) %>%
  mutate(id1 = rownames(juvie_mat)[as.integer(offspring)],
         id2 = rownames(adult_mat)[ind])

result %>% filter(number <= 1000)

```

Try plotting all those pairs, and compare them to the simulated values
```{r plot-all-pairs-density, fig.width = 12}
simlogls <- extract_logls(Q = Qs, numer = c(PO = 1), denom = c(U = 1))
ggplot(result, aes(x = value)) +
  geom_density(colour = "red") +
  geom_density(data = simlogls, aes(x = logl_ratio, fill = true_relat), alpha = 0.25) +
  scale_x_continuous(breaks = seq(-170, 60, by = 10))
```
OK, that is nice to see.  It is a little hard to make out---the red line above is the
observed distribution from all the pairs.  You see that it lands almost perfectly on
top of the simulated distribution for unrelated pairs.  That is good! It looks super clean.  


However, there are some troubling things here.  With a Logl Cutoff of 10 we should
expect almost no false positives, and yet R031681 is in three different pairs at that
level.  That is the only one that seems to have multiple parents, but it is a little
bit disquieting.  Here are this individual's genotypes:
```{r weird-fish}
 kelp_idx_frame %>% 
  filter(NMFS_DNA_ID == "R031681")
```
And here we can look at all its info:
```{r all-weird}
kelp_genos %>% 
  filter(NMFS_DNA_ID == "R031681") 
```
It has good read depths, but terrible allele balances at most loci.  That is super fishy.

What if we look at one of the other fish in there?
```{r another-one}
kelp_genos %>% 
  filter(NMFS_DNA_ID == "R015337") 
```
Those allele balances all tend to be larger than 0.7

### Contamination at R031681?

OK! My guess is that R031681 is a mixture of DNA from a two or more fish.  Since we are just taking the top two highest read-depth
haplotypes from each individual, at present, we wouldn't notice that this individual had super high read depths.  But,
we can easily go figure that out by grabbing this guys complete raw data and having a look at it.  
```{r check-contam, rows.print=20}
readRDS("../extdata/haplot-rds/gtseq50_96loci.rds") %>%
  tbl_df() %>%
  filter(id == "s13413") %>%
  arrange(locus, desc(depth))
```
Let's compare that to another individual, our friend R015337 from above:
```{r check-contam2, rows.print=20}
readRDS("../extdata/haplot-rds/gtseq33_96loci.rds") %>%
  tbl_df() %>%
  filter(id == "s7267") %>%
  arrange(locus, desc(depth))
```
Wow! Take a look at Plate_2_G06_Sat_EW987118_consensus (tab 3) between those fish, or tag_id_1184.  

Another thing to look at would be to check out the individual heterozygosities of all these fish.

### Looking at internal heterozygosities

If that guy is noticeably contaminated, we should see it in elevated internal heterozygosities.
We can do that like this:
```{r compute-int-hzs}
int_hz <- kg2 %>% 
  group_by(NMFS_DNA_ID, Locus) %>% 
  summarise(hz = Allele[1]!=Allele[2]) %>% 
  filter(!is.na(hz)) %>% 
  group_by(NMFS_DNA_ID) %>% 
  summarise(num_loc = n(), num_hz = sum(hz), fract_hz = num_hz / num_loc)
```
And now let us plot that and also put all the individuals (parents and offpring) that look likely
above on it.
```{r plot-int-hz, fig.width=12}
# first get the IDs of all the ones in pairs with LogL > 10
tmp <- result %>% filter(value > 10)
likelies <- int_hz %>% 
  filter(NMFS_DNA_ID %in% unique(c(tmp$id1, tmp$id2))) %>%
  arrange(fract_hz) %>%
  mutate(y = -0.1)

ggplot(int_hz, aes(x = fract_hz)) +
  geom_density(fill = "orange", alpha = 0.3) +
  geom_point(data = likelies, mapping = aes(x = fract_hz, y = y)) +
  geom_text(data = likelies, mapping = aes(x = fract_hz, y = y, label = NMFS_DNA_ID), angle = 60, hjust = 1.2, size = 3) +
  ylim(-1,9)
```
OK! That is pretty definitive: R031681 is out!


## Checking the logls

As long as we are at it, we may as well plot the distribution of our LogL>10, non-contaminated individuals 
against the backdrop of our simulated PO/U LogL values for true parent offspring pairs, to make sure there is
nothing weird going on there.
```{r plot-po-logls}
good_uns <- result %>% 
  filter(value > 10,
         id1 != "R031681") %>%
  mutate(y = 0.001)
po_frame <- simlogls %>% 
  filter(true_relat == "PO")
ggplot() +
  geom_density(data = po_frame, aes(x = logl_ratio), fill = "green", alpha = 0.3) +
  geom_point(data = good_uns, aes(x = value, y = y))

```

Note that the two with the lowest LogLs are kinda lower than we would really expect, but
not totally impossible.  Check which ones those are again:
```{r print-good-uns}
good_uns
```
But, otherwise, those values look pretty reasonable.  

## Could they be other types of relatives?

We have seen above that most full siblings  can have PO/U logls that are right around 10.  Half-siblings can also
see PO/U logls in that ballpark.  It is super unlikely that you would have an adult full sibling of anyone recruiting
this year, but it is possible that you could have a half-sibling.  I should look at the PO/HS and PO/FS LogLs for all
8 of our candidate PO pairs.  First off, let's see what those ratios look like for 
PO vs HS or FS.  
```{r plot-fs-ratios, fig.width = 8}
po_fs <- extract_logls(Qs, numer = c(PO = 1), denom = c(FS = 1)) 
ggplot(po_fs, aes(x = logl_ratio, fill = true_relat)) +
  geom_density(alpha = 0.25) +
  ggtitle("PO/FS Log-likelihood Ratios")
```

```{r plot-hs-ratios, fig.width = 8}
po_hs <- extract_logls(Qs, numer = c(PO = 1), denom = c(HS = 1)) 
ggplot(po_hs, aes(x = logl_ratio, fill = true_relat)) +
  geom_density(alpha = 0.25) +
  ggtitle("PO/HS Log-likelihood Ratios")
```

OK, so, not super easy to resolve these things, but it might be worth computing them.

### Computing PO/FS and PO/HS for the good_uns

First we need to filter down our genotype matrices to include only the offspring and the adults that
are in good_uns.
```{r filter-matrices}
adult_mat_filt <- adult_mat[good_uns$id2, ]
juvie_mat_filt <- juvie_mat[good_uns$id1, ]
```
Then flatten some more CKMR objects and compute the ratios:
```{r more-flattening}
fs_flat <- flatten_ckmr(CK, "FS")
hs_flat <- flatten_ckmr(CK, "HS")

po_fs_logl_flat <- po_flat
po_fs_logl_flat$probs <- log(po_flat$probs / fs_flat$probs)

po_hs_logl_flat <- po_flat
po_hs_logl_flat$probs <- log(po_flat$probs / hs_flat$probs)
```

Then do the comparisons.  This is fast because it is only a few of them.
```{r do-handful-of-comps-fs}
idx <- 1:nrow(juvie_mat_filt)
names(idx) <- idx
result_po_fs <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = adult_mat_filt, T = juvie_mat_filt, t = i, values = po_fs_logl_flat$probs, nGenos = po_fs_logl_flat$nGenos, Starts = po_fs_logl_flat$base0_locus_starts)
    #tmp[rev(top_index(tmp$value, 5)), ]  # just take the top 5 from each
    }) %>%
  dplyr::bind_rows(.id = "offspring") %>%
  dplyr::tbl_df()  %>%
  filter(offspring == ind) %>%  # just filter it down to the ones we want here...
  rename(po_fs_logl = value) %>%
  mutate(id1 = rownames(juvie_mat_filt)[as.integer(offspring)],
         id2 = rownames(adult_mat_filt)[ind])

result_po_fs
```

And the same for half-sibling ratios:
```{r do-handful-of-comps-hs}
idx <- 1:nrow(juvie_mat_filt)
names(idx) <- idx
result_po_hs <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = adult_mat_filt, T = juvie_mat_filt, t = i, values = po_hs_logl_flat$probs, nGenos = po_hs_logl_flat$nGenos, Starts = po_hs_logl_flat$base0_locus_starts)
    #tmp[rev(top_index(tmp$value, 5)), ]  # just take the top 5 from each
    }) %>%
  dplyr::bind_rows(.id = "offspring") %>%
  dplyr::tbl_df()  %>%
  filter(offspring == ind) %>%  # just filter it down to the ones we want here...
  rename(po_hs_logl = value) %>%
  mutate(id1 = rownames(juvie_mat_filt)[as.integer(offspring)],
         id2 = rownames(adult_mat_filt)[ind])

result_po_hs
```
So, R013607 is not looking super strong as a PO pair.  Could be a half-sibling.  There are so few of them that it
seems like we might want to consider confirming them with a ddRAD run or something.

All in all though, I think we do have a handful of very strong PO pairs.

## Look at the meta data for those fish

Have a quick looksie:
```{r po-meta}
meta_trim <- meta %>%
  select(NMFS_DNA_ID, SAMPLE_ID, LENGTH, REPORTED_LIFE_STAGE, COLLECTION_DATE, SITE, LATITUDE_M, LONGITUDE_M, SAMPLE_COMMENTS, LOCATION_COMMENTS_M)

good_pairs <- good_uns %>%
  select(-offspring, -ind, -y, -number) %>%
  rename(offspring = id1, 
         parent = id2,
         po_u_logl = value) %>%
  select(offspring, parent, everything())

meta_juv <- meta_trim
names(meta_juv) <- paste("juv", names(meta_juv), sep = "_")
meta_ad <- meta_trim
names(meta_ad) <- paste("ad", names(meta_ad), sep = "_")

pairs_with_meta <- left_join(good_pairs, meta_juv, by = c("offspring" = "juv_NMFS_DNA_ID")) %>%
  left_join(., meta_ad, by = c("parent" = "ad_NMFS_DNA_ID"))

write_csv(pairs_with_meta, "csv_outputs/pairs_with_meta.csv")

pairs_with_meta
```



## Let's look for full-siblings

### Power full full-sib detection

First, we want to see what the distributions of FS/U LogLs look like for
FS, U, and HS.  This is pretty straightforward to do with the ckmr object we made:
```{r plot-fs-logls, fig.width=8}
fs_u <- extract_logls(Qs, numer = c(FS = 1), denom = c(U = 1))
ggplot(fs_u, aes(x = logl_ratio, fill = true_relat)) +
  geom_density(alpha = 0.25) +
  ggtitle("FS/U Log-likelihood Ratios")
```

And we can compute the false positive rates as well:
```{r fs-fprs}
mc_sample_simple(Qs, nu = "FS", de = c("U"), tr = c("U", "FS", "HS", "PO"), method = "both")
```

### Doing the pairwise comparisons

I am just going to bung the parents and juveniles together into a single genotype matrix of all the kelps
and then I will use my existing function with the same matrix as source and target, then filter out half of 
them.
```{r bung-compare-filter}
all_kelp_pre <- rbind(adult_mat, juvie_mat)

# keep only those with < 0.61 internal HZ
hz_keepers <- int_hz$NMFS_DNA_ID[int_hz$fract_hz < 0.61]
hz_keepers <- intersect(hz_keepers, rownames(all_kelp_pre))
all_kelp_mat <- all_kelp_pre[hz_keepers, ]

# then make the flattened thing
fs_u_flat <- fs_flat
fs_u_flat$probs <- log(fs_flat$probs / unrel_flat$probs)

# then do the comparisons, only keep the top 20 for each one
idx <- 1:nrow(all_kelp_mat)
names(idx) <- idx
fs_comps <- lapply(idx, function(i) {
    tmp <- comp_ind_pairwise(S = all_kelp_mat, T = all_kelp_mat, t = i, values = fs_u_flat$probs, nGenos = fs_u_flat$nGenos, Starts = fs_u_flat$base0_locus_starts)
    tmp[rev(top_index(tmp$value, 20)), ]
    }) %>%
  dplyr::bind_rows(.id = "member1") %>%
  dplyr::tbl_df() %>%
  rename(member2 = ind) %>%
  mutate(member1 = as.integer(member1)) %>%
  filter(member1 < member2) %>%
  mutate(name1 = rownames(all_kelp_mat)[member1],
         name2 = rownames(all_kelp_mat)[member2]) %>%
  arrange(desc(value))

```

OK, that is pretty cool.  But we really need a little metadata on there to know who these individuals are.
```{r tidy-fs-comps-meta}
mini_meta <- meta %>%
  select(NMFS_DNA_ID, REPORTED_LIFE_STAGE, COLLECTION_DATE, LATITUDE_M, LONGITUDE_M)
mm1 <- mini_meta
names(mm1) <- c("name1", "stage1", "date1", "lat1", "long1")
mm2 <- mini_meta
names(mm2) <- c("name2", "stage2", "date2", "lat2", "long2")

fs_comps_meta <- fs_comps %>%
  left_join(., mm1) %>%
  left_join(., mm2)

# and, don't forget that we have to toss out the parent-offspring pairs 
# that we found
dumpers <- data_frame(name1 = c(good_uns$id1, good_uns$id2),
                      name2 = c(good_uns$id2, good_uns$id1))
  
high_likely_fs <- fs_comps_meta %>%
  filter(value > 13) %>%
  anti_join(dumpers) %>%
  arrange(member1, member2)
  
high_likely_fs


```
Holy Guacomole!  I think what we have here is clear evidence of at least one 
large full sibling groups amongst the adults.  Like more than 15 individuals. Wow!  Also,
some full siblings amongst the juveniles---but not massive full sib groups.  

At a cursory glance it appears that all the 
dates check out.  

I need to find connected components to figure out what the actual FS groups are.
We do this with some graph libraries
```{r find-connected-groups}
edgelist <- lapply(1:nrow(high_likely_fs), function(x) c(high_likely_fs$name1[x], high_likely_fs$name2[x]))
gR <- gRbase::ug(edgelist)
CnC <- graph::connComp(gR)  # these are the connected components
CnC_sort <- CnC[base::order(sapply(CnC, length), decreasing = TRUE)]

# so then let's make a sibgroup column in 
# the high_likely_fs frame
sib_group_pairs <- lapply(1:length(CnC), function(i) {
  high_likely_fs %>%
    filter(name1 %in% CnC_sort[[i]] | name2 %in% CnC[[i]])
}) %>%
  bind_rows(.id = "sibgroup") %>%
  mutate(sibgroup = as.integer(sibgroup)) %>%
  arrange(sibgroup, desc(value))

# let's save that as CSV
write_csv(sib_group_pairs, path = "csv_outputs/kelp_rockfish_sibgroup_pairs.csv")

# and show it here
sib_group_pairs
```

So, some things to note.  That appears to be a full sibling group of 19 adults.
Also, it looks to me like `R014981` and `R014982` in sibgroup 2 are probably duplicate
samples of some sort that I did not manage to filter out. Or there is something
going on with distortion of allele frequencies

### Is this really a big sibgroup?  

So, this is exciting.  But we need to confirm that this really looks like a large sibgroup.


One way we can do that is by counting up the number and distribution of alleles within 
it.  Obviously there should be no more than 4 alleles and they should be in frequencies of
multiples of 1/4 or 1/2.  We could do even more by making sure the genotypes were feasible
FSG-graph genotypes, but we will save that for next (or later).
```{r check-sibgroup-alleles}
big_sibgroup_genos <- left_join(kg2, kg_ckmr_markers) %>%
  filter(NMFS_DNA_ID %in% CnC_sort[[1]])

# now tally up the number of occurrences of each allele at each locus
big_sibgroup_genos %>%
  group_by(Locus, Pos, Allele, LocIdx, AlleIdx) %>%
  tally() %>%
  ungroup() %>%
  arrange(LocIdx, desc(n))
  
```
That is not reassuring.  The patterns there do not look like what we would have with a single large full sibling group
unless there was rampant genotyping error.  

So, let's see what happens if we only take super high-likelihood full sib pairs...
```{r make-super-high-logl-pairs}
super_high_likely_fs <- fs_comps_meta %>%
  filter(value > 35) %>%
  anti_join(dumpers) %>%
  arrange(member1, member2)

edgelist2 <- lapply(1:nrow(super_high_likely_fs), function(x) c(super_high_likely_fs$name1[x], super_high_likely_fs$name2[x]))
gR2 <- gRbase::ug(edgelist2)
high_CnC <- graph::connComp(gR2)  # these are the connected components
high_CnC_sort <- high_CnC[base::order(sapply(high_CnC, length), decreasing = TRUE)]

stricter_big_sibgroup_genos <- left_join(kg2, kg_ckmr_markers) %>%
  filter(NMFS_DNA_ID %in% high_CnC_sort[[1]])

# now tally up the number of occurrences of each allele at each locus
stricter_big_sibgroup_genos %>%
  group_by(Locus, Pos, Allele, LocIdx, AlleIdx) %>%
  tally() %>%
  ungroup() %>%
  arrange(LocIdx, desc(n))

```

That is a situation in which we only have retained individuals in the sibgroup which have
a full-sib loglikelihood ratio of more than 35 (which is huge) to at least one other 
individual in the full sibship.  So, these should all certainly be full siblings
in a perfect world.  There are 15 of those individuals.  So, 30 gene copies that we are playing with.  
This still doesn't look like an actual full sibship.

#### Going for the genotype graphs

OK, we had best look at the genotypes themselves a la Almudevar's 2000 or so paper.
I am just going to tally up allele-pairs in sorted order and that should give us
even more insight into this, I hope.

I am also going to slap on the population allele frequencies of these alleles so
we can see what is going on.  

```{r fsg-geno-graphs}
afreq_tmp1 <- kg_ckmr_markers %>%
  select(LocIdx, AlleIdx, Freq)
afreq_tmp2 <- afreq_tmp1 
names(afreq_tmp1) <- c("LocIdx", "alle1", "popfreq1")
names(afreq_tmp2) <- c("LocIdx", "alle2", "popfreq2")

stricter_big_sibgroup_genos %>%
  group_by(NMFS_DNA_ID, Locus, LocIdx) %>%
  summarise(alle1 = sort(AlleIdx)[1],
            alle2 = sort(AlleIdx)[2],
            geno = paste(alle1, alle2, sep = "-")) %>%
  group_by(Locus, LocIdx, alle1, alle2, geno) %>%
  tally() %>%
  ungroup() %>%
  left_join(afreq_tmp1) %>%
  left_join(afreq_tmp2) %>%
  arrange(LocIdx, desc(n)) %>%
  mutate(pop_geno_prob = popfreq1 * popfreq2 * (1 + (alle1 != alle2)))
```
So, I'm not quite sure what to make of this at this point.

These are the same sorts of patterns I discerned in my brief foray into Dani Ottman's data: a lot of pairs that looked
like they must be full siblings---but when you look at them all together it is clear that they cannot be a single, large full
sibling group.

I have already tossed out any individuals that had low internal heterozygosity, so I don't think that is driving the pattern.

The fact that all the siblings in that group were adults is somewhat reassuring that it might not be completely artifactual.  It
could have something to do with a particular cohort/year-class.

### Locus Specific Contributions To These LogLs

For each pair amongst the super-hi LogL sibship, I wish to investigate the distribution of locus contributions to these likelhoods.
First, we get all the pairs that have LogLs above 35 and in which one member is in the 
large connected group.
```{r get-super-hi-likelies}
super_hi_pairs1 <- super_high_likely_fs %>%
  filter(name1 %in% high_CnC_sort[[1]] | name2 %in% high_CnC_sort[[1]])
```
There are only 24 of those, which tells us something: very few of the pairs in this connected group
actually have high full-sib LogL values. OK.

Now we have to make genotype matrices for name1 and name2 members and look at the locus-specific
contributions:
```{r loc-spec-contrib}
SS <- all_kelp_mat[super_hi_pairs1$name1,]
TT <- all_kelp_mat[super_hi_pairs1$name2,]

loc_spec_vals <- locus_specific_pairwise(S = SS, T = TT,  s = 1:nrow(SS),  t = 1:nrow(TT), values = fs_u_flat$probs, nGenos = fs_u_flat$nGenos, Starts = fs_u_flat$base0_locus_starts) %>%
  tbl_df() %>%
  group_by(locus) %>%
  mutate(locus_mean_val = mean(value)) %>%
  arrange(desc(locus_mean_val))

loc_spec_vals
```
We can see which loci on average are contributing a lot to the LogL and then go check the 
genotype counts at those loci.  There are some funny things, it seems to me. I'm perplexed. 
I suppose it could be a bunch of nested half-sibs.  It is odd.  

## Dump those dudes to Colony

I am going to save the IDs of those individuals that were in the big group
that looked sibship-like, and I will dump their genotypes as well and the 
allele freqs for all the alleles so that we can use that in Colony.  

So, what we do here is dump a bunch of stuff
```{r assemble-output-list}

duplicate_samples <- def_same %>% 
  tbl_df() %>% 
  mutate(id1 = rownames(rocky_mat)[ind1], id2 = rownames(rocky_mat)[ind2])


output_list <- list(
  kelp_idx_frame = kelp_idx_frame,  # these are the haplotypes of each individual in numeric-allelic form
  kg_ckmr_markers = kg_ckmr_markers,   # allele freqs in CKMR long format
  CnC_sort = CnC_sort,
  top_po_pairs = result %>% filter(number <= 11),
  duplicate_samples = duplicate_samples,
  int_hz = int_hz
)

dir.create("rds_outputs")
saveRDS(output_list, file = "rds_outputs/kr-parentage-output-list.rds", compress = "xz")

```